chalk-solve/src/clauses/builtin_traits/fn_family.rs

```diff
@@ -19,11 +65,71 @@ use chalk_ir::{
 pub fn add_fn_trait_program_clauses<I: Interner>(
     db: &dyn RustIrDatabase<I>,
     builder: &mut ClauseBuilder<'_, I>,
-    trait_id: TraitId<I>,
+    well_known: WellKnownTrait,
     self_ty: Ty<I>,
 ) -> Result<(), Floundered> {
     let interner = db.interner();
+    let trait_id = db.well_known_trait_id(well_known).unwrap();
+
     match self_ty.data(interner) {
+        TyData::Apply(apply) => match apply.name {
+            TypeName::FnDef(fn_def_id) => {
+                let fn_def_datum = builder.db.fn_def_datum(fn_def_id);
+                let bound = fn_def_datum
+                    .binders
+                    .substitute(builder.interner(), &apply.substitution);
+                let inputs_and_output = &bound.inputs_and_output;
+                builder.push_binders(inputs_and_output, |builder, inputs_and_output| {
+                    let self_ty = ApplicationTy {
+                        name: apply.name,
+                        substitution: builder.substitution_in_scope(),
+                    }
+                    .intern(interner);
+
+                    let arg_sub = inputs_and_output
+                        .argument_types
+                        .iter()
+                        .cloned()
+                        .map(|ty| ty.cast(interner));
+                    let arg_sub = Substitution::from(interner, arg_sub);
+                    let output_ty = inputs_and_output.return_type;
+
+                    push_clauses(
+                        db, builder, well_known, trait_id, self_ty, arg_sub, output_ty,
+                    );
+                });
+                Ok(())
+            }
+            TypeName::Closure(closure_id) => {
+                let closure_datum = db.closure_datum(closure_id, &apply.substitution);
+                let trait_matches = match (well_known, closure_datum.kind) {
+                    (WellKnownTrait::Fn, ClosureKind::Fn) => true,
+                    (WellKnownTrait::FnMut, ClosureKind::FnMut)
+                    | (WellKnownTrait::FnMut, ClosureKind::Fn) => true,
+                    (WellKnownTrait::FnOnce, _) => true,
+                    _ => false,
+                };
+                if !trait_matches {
+                    return Ok(());
+                }
+                let inputs_and_output = &closure_datum.inputs_and_output;
+                builder.push_binders(inputs_and_output, |builder, inputs_and_output| {
+                    let arg_sub = inputs_and_output
```

Comment:
Nit: it feels like the construction of `arg_sub` and `output_ty` could be moved into the helper routine as well..? I think they are identical

Reply:
Ah, I see, there's a third call. I guess then there could be a second helper.

Reply:
But it's ok as is.

Reply:
I went ahead and added a helper function for the fndef and closure case.